### ISSUE TITLE
Deployment fails at the "Download Artifacts" stage

### ISSUE THREAD

**Issue by akunaatrium** [Microsoft/vsts-tasks#3266]
Here is the deployment process log:
[log.txt](https://github.com/Microsoft/vsts-tasks/files/657232/log.txt)

**Comment by DavidStaheli** [comment]
`2016-12-16T12:24:07.9833191Z ##[warning]Re-attempting download of C:\a\r1\a\CI Build\drop\PackagingPublishingWorkerService.zip.download. Error: The read operation failed, see inner exception.`

This may have been caused by a live site issue in the Packaging Service earlier today.

**Comment by bryanmacfarlane** [comment]
@DavidStaheli - our artifacts don't use packaging service yet.  RM owns the agent plugin to get artifacts.  We need to ensure it has retries etc.. for robustness which we did on the upload side.

**Comment by bryanmacfarlane** [comment]
So - this should be an agent issue ...

**Comment by akunaatrium** [comment]
The error doesn't seem to be happening any more btw.

**Comment by GitHubSriramB** [comment]
@akunaatrium - thanks for sharing the logs.

We do have retries in place when downloading artifacts. However we had a bug due to which we didn't stop the download cleanly before retrying. This is what has caused the download to fail. We have fixed this issue recently. Hopefully retrying download should help now.

Let us know if you hit this issue again.

Analysis from logs:

We retried downloading one of the items 5 times, but it didn't help:
…
2016-12-16T12:22:47.8166848Z ##[warning]Re-attempting download of C:\a\r1\a\CI Build\drop\PackagingPublishingWorkerService.zip.download. Error: The read operation failed, see inner exception.
…
2016-12-16T12:33:19.6896931Z ##[warning]Re-attempting download of C:\a\r1\a\CI Build\drop\Ves.WebApi\obj\Release\Package\Ves.WebApi.zip.download. Error: The read operation failed, see inner exception.
…
2016-12-16T12:33:19.7984372Z An error occurred during download: System.Exception: Error downloading C:\a\r1\a\CI Build\drop\PackagingPublishingWorkerService.zip.download: System.IO.IOException: The read operation failed, see inner exception. ---> System.Net.Http.WinHttpException: The connection with the server was terminated abnormally

After this, we retried downloading the entire artifact again. But then we hit an error due to sharing violation (this is due to a bug where we didn’t cancel the ongoing downloads cleanly):
…
2016-12-16T12:33:19.7994370Z Retrying download...
2016-12-16T12:33:19.8034365Z Ensuring artifact folder C:\a\r1\a\CI Build exists and is clean.
2016-12-16T12:33:19.9124369Z An error occurred during download: System.AggregateException: One or more errors occurred. (One or more errors occurred. (The process cannot access the file 'C:\a\r1\a\CI Build\drop\Ves.Web\obj\Release\Package\Ves.Web.zip.download' because it is being used by another process.)) (The process cannot access the file 'C:\a\r1\a\CI Build\drop\Ves.Web\obj\Release\Package\Ves.Web.zip.download' because it is being used by another process.) ---> System.IO.IOException: The process cannot access the file 'C:\a\r1\a\CI Build\drop\Ves.Web\obj\Release\Package\Ves.Web.zip.download' because it is being used by another process.
…
